File: python/stereoBM_Grad.py
# Stereo Matching using Block Matching (Image Gradients)
# Computes a disparity map from a rectified stereo pair using Block Matching

import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt
from shiftArray import shiftArray

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
windowSize = 5

# Define data cost computation
dataCostComputation = lambda leftX,leftY,rightX,rightY: np.absolute(leftX-rightX)+np.absolute(leftY-rightY) #magnitude

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Compute image gradients
leftGradX = cv.Sobel(leftImg,cv.CV_64F,1,0,ksize=3)
leftGradY = cv.Sobel(leftImg,cv.CV_64F,0,1,ksize=3)
rightGradX = cv.Sobel(rightImg,cv.CV_64F,1,0,ksize=3)
rightGradY = cv.Sobel(rightImg,cv.CV_64F,0,1,ksize=3)

# Compute pixel-based matching cost (data cost)
dataCost = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    rightGradXShifted = shiftArray(rightGradX,[0,d])
    rightGradYShifted = shiftArray(rightGradY,[0,d])
    #rightGradXShifted = np.roll(rightGradX,d,1) #less accurate, better performances
    #rightGradYShifted = np.roll(rightGradY,d,1) #less accurate, better performances
    dataCost[:,:,d] = dataCostComputation(leftGradX,leftGradY,rightGradXShifted,rightGradYShifted)

# Aggregate the matching cost
dataCost = cv.boxFilter(dataCost,-1,(windowSize,windowSize),normalize=False)

# Compute the disparity map
dispMap = np.argmin(dataCost,axis=2)

# Normalize the disparity map for display
scaleFactor = 256/dispLevels
dispImg = (dispMap*scaleFactor).astype(np.uint8)

# Show disparity map
plt.imshow(dispImg,cmap="gray")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparityBM_Grad.png",dispImg)

plt.show()

File: matlab/stereoBM_Grad.m
% Stereo Matching using Block Matching (Image Gradients)
% Computes a disparity map from a rectified stereo pair using Block Matching

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
windowSize = 5;

% Define data cost computation
dataCostComputation = @(left,right) sum(abs(left-right),3); %magnitude

% Start timer
timerVal = tic();

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Compute image gradients
leftGrad = zeros(rows,cols,2,'double');
[leftGrad(:,:,1),leftGrad(:,:,2)] = imgradientxy(leftImg,'sobel');
rightGrad = zeros(rows,cols,2,'double');
[rightGrad(:,:,1),rightGrad(:,:,2)] = imgradientxy(rightImg,'sobel');

% Compute pixel-based matching cost (data cost)
dataCost = zeros(rows,cols,dispLevels,'double');
for d = 0:dispLevels-1
    %rightGradShifted = shiftArray(rightGrad,[0,d,0]);
    rightGradShifted = circshift(rightGrad,d,2); %less accurate, better performances
    dataCost(:,:,d+1) = dataCostComputation(leftGrad,rightGradShifted);
end

% Aggregate the matching cost
dataCost = int32(convn(dataCost,ones(windowSize,windowSize,1),'same'));

% Compute the disparity map
[~,ind] = min(dataCost,[],3);
dispMap = ind-1;

% Normalize the disparity map for display
scaleFactor = 256/dispLevels;
dispImg = uint8(dispMap*scaleFactor);

% Show disparity map
figure; imshow(dispImg)

% Save disparity map
imwrite(dispImg,'disparityBM_Grad.png')

% Stop timer and display running time
elapsedTime = toc(timerVal);
fprintf('Running time: %.2f seconds\n',elapsedTime)
